<?xml version="2.0" encoding="UTF-8"?>
<launch>

  <!-- Map Server Arguments -->
  <arg name="map_file" default="$(find monte_carlo)/maps/map2.yaml"/>
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" />
  
  <!-- RViz Launching -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>

  <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
    <arg name="model" value="$(arg model)"/>
  </include>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find motion_planning)/rviz/motion_rviz.rviz"/>

  <!-- Movement Smoother -->


  <!-- Motion Planning -->

  <arg name="odom_frame_id"   default="odom"/>
  <arg name="base_frame_id"   default="base_footprint"/>
  <arg name="global_frame_id"   default="map"/>
  <arg name="odom_topic"   default="odom"/>
  <arg name="laser_topic"   default="scan"/>

  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen" clear_params="true">

    <rosparam file="$(find motion_planning)/config/costmap_common_params.yaml" command="load" ns="global_costmap"/>
    <rosparam file="$(find motion_planning)/config/costmap_common_params.yaml" command="load" ns="local_costmap"/>
    <rosparam file="$(find motion_planning)/config/local_costmap_params.yaml" command="load"/>
    <rosparam file="$(find motion_planning)/config/global_costmap_params.yaml" command="load"/>
    <rosparam file="$(find motion_planning)/config/base_local_planner_params.yaml" command="load"/>
    <rosparam file="$(find motion_planning)/config/navfn_global_planner_params.yaml" command="load"/>
    <rosparam file="$(find motion_planning)/config/recovery_behaviours.yaml" command="load"/>
  </node>

  <!-- AMCL -->
  <arg name="scan_topic"     default="scan"/>
  <arg name="initial_pose_x" default="0.0"/>
  <arg name="initial_pose_y" default="0.0"/>
  <arg name="initial_pose_a" default="0.0"/>

  <node pkg="amcl" type="amcl" name="amcl">

    <param name="transform_tolerance"       value="0"/>
    <param name="min_particles"             value="100"/>
    <param name="max_particles"             value="10000"/>
    <param name="kld_err"                   value="0.1"/>
    <param name="kld_z"                   value="0.99"/>
    <param name="update_min_d"              value="0.20"/>
    <param name="update_min_a"              value="0.70"/>
    <param name="resample_interval"         value="2"/>
    <!--<param name="transform_tolerance"       value="0.5"/> -->
    <param name="recovery_alpha_slow"       value="0.0001"/>
    <param name="recovery_alpha_fast"       value="0.1"/>
    <param name="initial_cov_xx"       value="15"/>
    <param name="initial_cov_yy"       value="15"/>
    <param name="initial_cov_aa"       value="15"/>
    <!--<param name="initial_pose_x"            value="$(arg initial_pose_x)"/>
    <param name="initial_pose_y"            value="$(arg initial_pose_y)"/>
    <param name="initial_pose_a"            value="$(arg initial_pose_a)"/>
    <param name="gui_publish_rate"          value="50.0"/>
    <remap from="scan"                      to="$(arg scan_topic)"/> -->
    <param name="laser_max_range"           value="-1.0"/>
    <!--<param name="laser_max_beams"           value="180"/>
    <param name="laser_z_hit"               value="0.5"/>
    <param name="laser_z_short"             value="0.05"/>
    <param name="laser_z_max"               value="0.05"/>
    <param name="laser_z_rand"              value="0.5"/>
    <param name="laser_sigma_hit"           value="0.2"/>
    <param name="laser_lambda_short"        value="0.1"/>
    <param name="laser_likelihood_max_dist" value="2.0"/>
    <param name="laser_model_type"          value="likelihood_field"/>
    <param name="odom_model_type"           value="diff"/>
    <param name="odom_alpha1"               value="0.1"/>
    <param name="odom_alpha2"               value="0.1"/>
    <param name="odom_alpha3"               value="0.1"/>
    <param name="odom_alpha4"               value="0.1"/>
    <param name="odom_frame_id"             value="odom"/>
    <param name="base_frame_id"             value="base_footprint"/> -->

  </node>

</launch>
